pkg/workloadlearning/handle.go

```diff
@@ -0,0 +1,213 @@
+//     from every record in statement_summary/statement_stats
+//
+// 2,3. Group by tablename, get the total scan time, total memory usage, and every table scan time, memory usage,
+//
+//	read frequency
+//
+// 4. Calculate table cost for each table, table cost = table scan time / total scan time + table mem usage / total mem usage
+// 5. Save all table cost metrics[per table](scan time, table cost, etc) to table "mysql.workload_values"
+func (handle *Handle) HandleReadTableCost(infoSchema infoschema.InfoSchema) {
+	// step1: abstract middle table cost metrics from every record in statement_summary
+	middleMetrics, startTime, endTime := handle.analyzeBasedOnStatementStats()
+	if len(middleMetrics) == 0 {
+		return
+	}
+	// step2: group by tablename, sum(table-scan-time), sum(table-mem-usage), sum(read-frequency)
+	// step3: calculate the total scan time and total memory usage
+	tableNameToMetrics := make(map[ast.CIStr]*ReadTableCostMetrics)
+	totalScanTime := 0.0
+	totalMemUsage := 0.0
+	for _, middleMetric := range middleMetrics {
+		metric, ok := tableNameToMetrics[middleMetric.TableName]
+		if !ok {
+			tableNameToMetrics[middleMetric.TableName] = middleMetric
+		} else {
+			metric.TableScanTime += middleMetric.TableScanTime * float64(middleMetric.ReadFrequency)
+			metric.TableMemUsage += middleMetric.TableMemUsage * float64(middleMetric.ReadFrequency)
+			metric.ReadFrequency += middleMetric.ReadFrequency
+		}
+		totalScanTime += middleMetric.TableScanTime
+		totalMemUsage += middleMetric.TableMemUsage
+	}
+	if totalScanTime == 0 || totalMemUsage == 0 {
+		return
+	}
+	// step4: calculate the percentage of scan time and memory usage for each table
+	for _, metric := range tableNameToMetrics {
+		metric.TableCost = metric.TableScanTime/totalScanTime + metric.TableMemUsage/totalMemUsage
+	}
+	// step5: save the table cost metrics to table "mysql.workload_values"
+	handle.SaveReadTableCostMetrics(tableNameToMetrics, startTime, endTime, infoSchema)
+}
+
+func (handle *Handle) analyzeBasedOnStatementSummary() []*ReadTableCostMetrics {
+	// step1: get all record from statement_summary
+	// step2: abstract table cost metrics from each record
+	return nil
+}
+
+// TODO
+func (handle *Handle) analyzeBasedOnStatementStats() ([]*ReadTableCostMetrics, time.Time, time.Time) {
+	// step1: get all record from statement_stats
+	// step2: abstract table cost metrics from each record
+	// TODO change the mock value
+	return nil, time.Now(), time.Now()
+}
+
+// SaveReadTableCostMetrics table cost metrics, workload-based start and end time, version,
+func (handle *Handle) SaveReadTableCostMetrics(metrics map[ast.CIStr]*ReadTableCostMetrics,
+	startTime, endTime time.Time, infoSchema infoschema.InfoSchema) {
+	// TODO save the workload job info such as start end time into workload_jobs table
+	// step1: create a new session, context, txn for saving table cost metrics
+	se, err := handle.sysSessionPool.Get()
+	if err != nil {
```

Comment:
we should defer before error check, otherwise, this is another goroutine leak

Reply:
The channel will return nil, false when sessionpool closed. If we put a nil into session pool, it will panic.
So we need to first check the error and then put the session pool